contracts/staking/ValidatorManager.sol

```diff
@@ -93,6 +90,104 @@ abstract contract ValidatorManager is
         $._pChainBlockchainID = settings.pChainBlockchainID;
         $._subnetID = settings.subnetID;
         $._maximumHourlyChurn = settings.maximumHourlyChurn;
+        $._initializedValidatorSet = false;
+    }
+
+    modifier initializedValidatorSet() {
+        require(
+            _getValidatorManagerStorage()._initializedValidatorSet,
+            "ValidatorManager: validator set not initialized"
+        );
+        _;
+    }
+
+    /**
+     * @notice See {IValidatorManager-initializeValidatorSet}.
+     */
+    function initializeValidatorSet(
+        SubnetConversionData calldata subnetConversionData,
+        uint32 messageIndex
+    ) external {
+        ValidatorManagerStorage storage $ = _getValidatorManagerStorage();
+        require(!$._initializedValidatorSet, "ValidatorManager: already initialized validator set");
+
+        WarpMessage memory warpMessage = _getPChainWarpMessage(messageIndex);
+        // Parse the Warp message into SubnetConversionMessage
+        bytes32 subnetConversionID =
+            ValidatorMessages.unpackSubnetConversionMessage(warpMessage.payload);
+
+        // Verify that the sha256 hash of the Subnet conversion data matches with the Warp message's subnetConversionID.
+        bytes memory encodedConversion = abi.encodePacked(
+            subnetConversionData.convertSubnetTxID,
+            subnetConversionData.blockchainID,
+            subnetConversionData.validatorManagerAddress
+        );
+        uint256 length = encodedConversion.length;
+        for (uint256 i; i < length; i++) {
+            InitialValidator memory initialValidator = subnetConversionData.initialValidators[i];
+            encodedConversion = abi.encodePacked(
+                encodedConversion,
+                initialValidator.nodeID,
+                initialValidator.weight,
+                initialValidator.blsPublicKey
+            );
+        }
+        require(
+            sha256(encodedConversion) == subnetConversionID,
+            "ValidatorManager: invalid subnet conversion ID"
+        );
+        require(
+            subnetConversionData.blockchainID == WARP_MESSENGER.getBlockchainID(),
+            "ValidatorManager: invalid blockchain ID"
+        );
+
+        // TODO: might have more efficient conversion with assembly, check length of bytes
+        require(
+            subnetConversionData.validatorManagerAddress.length == 20,
+            "ValidatorManager: invalid validator manager address length"
+        );
+        require(
+            address(bytes20(subnetConversionData.validatorManagerAddress)) == address(this),
+            "ValidatorManager: invalid validator manager address"
+        );
+        require(
+            subnetConversionData.initialValidators.length > 0,
+            "ValidatorManager: no initial validators"
+        );
+
+        // Insert the initial validators into the storage.
+        for (uint256 i; i < length; i++) {
```

Comment:
This `length` is not at all the right length to be using here

Reply:
as in how it was set right? Updated to be `subnetConversionData.initialValidators.length`

Reply:
Yeah, that's right